accepted/2020/better-obsoletion/securestring-obsoletion.md

```diff
@@ -0,0 +1,221 @@
+
+### Best practices for implementation
+
+If the backing buffer for the sensitive data is stored in-proc, we should attempt to use a separate unmanaged heap for storing these values, and we should not hand out direct references to this heap. This helps mitigate the possibility of sensitive data being leaked due to corruption of the normal heap. This concept is [already in use by OpenSSL](https://www.openssl.org/docs/man1.1.0/man3/CRYPTO_secure_malloc_initialized.html) and a few high-sensitivity Windows services. If using a separate heap is not feasible, we should attempt to use .NET's pinned object heap.
+
+The backing buffer may be (but is not required to be) enciphered or marked do-not-dump. In practice, we will probably encipher data when we can to avoid accidental leakage of the sensitive data. We will probably not mark the pages as do-not-dump due to the difficulty of bookkeeping and the fact that both Linux and Windows primarily intend do-not-dump regions to be for large data sections (such as texture data for games), and they institute an upper bound on the number of pages which may be annotated as do-not-dump.
+
+## New overloads on existing APIs
+
+Some APIs - particularly those dealing with cryptographic keys - are ideal candidates for being enlightened about `SensitiveData<T>`. A non-exhaustive list is provided below.
+
+```cs
+namespace System.Security.Cryptography
+{
+    public sealed class AesGcm
+    {
+        /* existing ctors */
+        public AesGcm(byte[] key);
+        public AesGcm(ReadOnlySpan<byte> key);
+
+        /* new proposed ctor */
+        public AesGcm(SensitiveData<byte> key);
+    }
+
+    public class Rfc2898DeriveBytes
+    {
+        /* existing method */
+        public static byte[] Pbkdf2(byte[] password, byte[] salt, int iterations, HashAlgorithmName hashAlgorithm, int outputLength);
+
+        /* new proposed overload - takes and returns a SensitiveData parameter */
+        public static SensitiveData<byte> Pbkdf2(SensitiveData<byte> password, byte[] salt, int iterations, HashAlgorithmName hashAlgorithm, int outputLength);
+    }
+}
+```
+
+## New proposed SecureString-replacement APIs
+
+For the X.509 APIs, we should expose APIs which take the password as a `SensitiveData<char>` or a `SensitiveData<byte>` directly. These would serve as overloads of the existing `SecureString`-consuming APIs.
+
+```cs
+namespace System.Security.Cryptography.X509Certificates
+{
+    public class X509Certificate
+    {
+        public X509Certificate(byte[] rawData, SensitiveData<char> password, X509KeyStorageFlags keyStorageFlags);
+        public virtual void Import(byte[] rawData, SensitiveData<char> password, X509KeyStorageFlags keyStorageFlags);
+        /* ... */
+    }
+}
+```
+
+For `Process.Start` and `ProcessStartInfo` we should expose overloads which take a `NetworkCredential`. These overloads would only work on Windows.
```

Comment:
Why only on Windows?

Reply:
This matches the behavior of the other password-consuming overloads today.